### PR TITLE
vSphere: zonal with external lb

### PR DESCRIPTION
**Pull request by jcpowermac**

### CODE REVIEW

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of this enhancement is to provide the ability to install vSphere IPI zonal on separate L2 segments. As of this writing the only way to 
+accomplish this is to allow an external load balancer.
+
+## Motivation
+
+Users of OpenShift would`` like the ability to deploy control plane and compute nodes in more than one L2 segment to increase reliability and ease of network management.
```
Comment (rvanderp3) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:46]: are the double backticks intentional?

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+### Non-Goals
+
+- Provide a mechanism to configure a external load balancer from the installer or machine-api-operator 
+
+## Proposal
+
+We propose the use of small L2 segments (e.g. /28). The customer's
+load balancer will be pre-configured with the entire subnet as end points
+for the required ports. Those L2 segments will be a associated with vSphere port groups that will be used in the `FailureDomains` and provisioning process.  The external load balancer will discover available endpoints and use those accordingly.
```
Comment (rvanderp3) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:61]: might link to the multi-zone enhancement

**Comment by jcpowermac** [comment]
fyi: @romfreiman @cybertron @patrickdillon @Miciah

**Comment by jcpowermac** [comment]
Currently working with haproxy and zonal, with changes to the installer. Obviously this isn't complete just a test.
https://github.com/jcpowermac/ocp-install-external-lb-hack

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of this enhancement is to provide the ability to install vSphere IPI zonal on separate L2 segments. As of this writing the only way to
```
Comment (cybertron) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:41]: Is this distinct from the remote workers functionality that we already have for baremetal? AFAIK that should already work for vsphere, we just haven't tested it.

https://issues.redhat.com/browse/OPNET-27

**Review by cybertron — COMMENTED** [review]
I know we have similar requests for other on-prem platforms (https://issues.redhat.com/browse/OPNET-14 ) so I'd be interested in a cross-platform solution to this. My thought was less removing all of the IPI components and making it behave like UPI, and more allow the DNS records to be pointed at an external loadbalancer.

That said I don't know all the details about your use cases. There are situations my idea would not support.

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md (hunk 3 above)
Comment (jcpowermac) [reply enhancements/installer/vsphere-ipi-zonal-external-lb.md:41]: ingress was resolved for vsphere
https://issues.redhat.com/browse/SPLAT-409

The scope is install time of a external lb without the static pods of IPI with support for api and ingress.

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of this enhancement is to provide the ability to install vSphere IPI zonal on separate L2 segments without the static pods (keepalived, haproxy, coredns).
```
Comment (romfreiman) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:41]: why without coredns?

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  // expected that the LB and DNS A records be pre-configured prior to installation.
+  // +kubebuilder:default=External
+  // +kubebuilder:validation:Enum=Internal;External
+  LoadBalancer LoadBalancerType `json:"loadBalancerType,omitempty"`
+}
+```
+
+### MCO
+
+The existing switch to deploy keepalived, haproxy and coredns static pods in the machine config operator is the apivip variable. This would change to be the Infrastructure status parameter: `LoadBalancerType`.
```
Comment (romfreiman) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:102]: for all on prem platforms or only for vsphere?
I would prefer to have it a generic enhancement to the on prem networking

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+const (
+	//LoadBalancerTypeExternal
+	LoadBalancerTypeExternal LoadBalancerType = "External"
+
+	// LoadBalancerTypeInternal
+	LoadBalancerTypeInternal LoadBalancerType = "Internal"
+)
+
+// Platform stores any global configuration used for vsphere platforms
+type Platform struct {
```
Comment (romfreiman) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:132]: why is this vpshere specific?

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 7]
```diff
@@ -0,0 +1,201 @@
+---
+title: vsphere-ipi-zonal-external-lb
```
Comment (romfreiman) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:2]: general thought:
The fact that there is a inter-dependency between IPI and onprem networking is a big limitation that we've been imposing on our customers. Hence I think having this enhancement is a great step forward. I do believe that it should be solved in a holistic manner - for all on prem platforms, rather than for vpshere only.
In assisted - the way we are categorizing it is - CMN (cluster managed netoworking) and UMN (user managed networking) - where, currently, it boils down to ipi/upi manifests (which define the networking yamls). Once this enhancement is implemented, we'll have to adjust our logic there.

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+authors:
+  - "@jcpowermac"
+reviewers:
+  - "@rvanderp3"
+  - "@bostrt"
```
Comment (romfreiman) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:7]: we need assisted installer team to chime in since it will affect vsphere integration of AI.
@oourfali @filanov @gamli75

File: enhancements/installer/vsphere-ipi-zonal-external-lb.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+- Modification to the Infrastructure status object to include new LoadBalancerType
+- Modification to the platform spec object to include new LoadBalancerType
+- Modification to the machine-config-operator to use new LoadBalancerType 
+
+### Workflow Description
+
+### api
+
+#### Infrastructure Status
```
Comment (romfreiman) [inline enhancements/installer/vsphere-ipi-zonal-external-lb.md:72]: what about existence / non-existence of VIPs - why it cannot be the signal? then it will be aligned with upi/ipi networking parts

Comment (romfreiman) [reply enhancements/installer/vsphere-ipi-zonal-external-lb.md:72]: and, the installer can set the lodbalancer type as the derivative of those values, without letting the user specifying it

**Comment by jcpowermac** [comment]
/hold

After a slack thread chat last week with @tsorya this enhancement will change significantly. 
tl;dr we can just change the IPI validation to allow no VIP(s) like UPI.

**Comment by jcpowermac** [comment]
/close
This PR from the installer already changed the VIP validation required to make a external load balancer for vSphere IPI zonal possible: https://github.com/openshift/installer/pull/5798
